include/core/ModelVariable.hpp

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+struct ModelVariable
+{
+    virtual ~ModelVariable() = default;
+
+    virtual shared_ptr<Discretization> discretization() const = 0;
+
+    size_t size() const { return this->discretization()->size(); }
+    int rank() const { return this->discretization()->rank(); }
+
+    udunits_dimension_t element_dimensionality();
+    udunits_units_t element_units();
+    
+    std::span<double> storage() const = 0;
+
+    std::string name() { return name_; }
+
+private:
+    std::string name_;
+};
+
+struct SourceVariable : public virtual ModelVariable
+{
+    shared_ptr<DataProvider> provider;
+
+};
+
+
+struct BmiSourceVariable : public virtual SourceVariable
+{
+    // constraint: dynamic_cast<Bmi_Formulation*>(provider) != nullptr
```

Comment:
This should try to use BMI's `get_value_ptr` to arrange `storage()`, and fall back to allocating a buffer that it will load into